src/code/z_lights.c

```diff
@@ -1,190 +1,381 @@
+    gSPLight(gfxCtx->polyXlu.p++, &lights->l.a, i);
+
+    Graph_CloseDisps(dispRefs, gfxCtx, "../z_lights.c", 352);
+}
 
-Light* Lights_MapperGetNextFreeSlot(LightMapper* mapper) {
-    if (6 < mapper->numLights) {
+Light* Lights_FindSlot(Lights* lights) {
+    if (lights->numLights >= 7) {
         return NULL;
+    } else {
+        return &lights->l.l[lights->numLights++];
     }
-
-    return &mapper->lights[mapper->numLights++];
 }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_lights/func_8007A0B4.s")
+void Lights_UpdatePoint(Lights* lights, LightParams* params, Vec3f* vec) {
+    f32 xDiff;
+    f32 yDiff;
+    f32 zDiff;
+    f32 posDiff;
+    f32 scale;
+    Light* light;
+
+    if (vec != NULL) {
+        xDiff = params->point.x - vec->x;
+        yDiff = params->point.y - vec->y;
+        zDiff = params->point.z - vec->z;
+        scale = params->point.radius;
+        posDiff = SQ(xDiff) + SQ(yDiff) + SQ(zDiff);
+
+        if (posDiff < SQ(scale)) {
+            light = Lights_FindSlot(lights);
+
+            if (light != NULL) {
+                posDiff = sqrtf(posDiff);
+                if (1) {}
+                scale = posDiff / scale;
+                scale = 1 - SQ(scale);
+
+                light->l.col[0] = light->l.colc[0] = params->point.color[0] * scale;
+                light->l.col[1] = light->l.colc[1] = params->point.color[1] * scale;
+                light->l.col[2] = light->l.colc[2] = params->point.color[2] * scale;
+
+                scale = (posDiff < 1.0f) ? 120.0f : 120.0f / posDiff;
+
+                light->l.dir[0] = xDiff * scale;
+                light->l.dir[1] = yDiff * scale;
+                light->l.dir[2] = zDiff * scale;
+            }
+        }
+    }
+}
 
-void func_8007A40C(LightMapper* mapper, LightInfoDirectionalParams* params, GlobalContext* globalCtx) {
-    Light* light = Lights_MapperGetNextFreeSlot(mapper);
+void Lights_UpdateDirectional(Lights* lights, LightParams* params, Vec3f* vec) {
+    Light* light = Lights_FindSlot(lights);
 
     if (light != NULL) {
-        light->l.col[0] = light->l.colc[0] = params->red;
-        light->l.col[1] = light->l.colc[1] = params->green;
-        light->l.col[2] = light->l.colc[2] = params->blue;
-        light->l.dir[0] = params->dirX;
-        light->l.dir[1] = params->dirY;
-        light->l.dir[2] = params->dirZ;
+        light->l.col[0] = light->l.colc[0] = params->dir.color[0];
+        light->l.col[1] = light->l.colc[1] = params->dir.color[1];
+        light->l.col[2] = light->l.colc[2] = params->dir.color[2];
+        light->l.dir[0] = params->dir.x;
+        light->l.dir[1] = params->dir.y;
+        light->l.dir[2] = params->dir.z;
     }
 }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_lights/func_8007A474.s")
+void Lights_Update(Lights* lights, LightNode* head, Vec3f* vec) {
+    LightsUpdateFunc updateFuncs[] = { Lights_UpdatePoint, Lights_UpdateDirectional, Lights_UpdatePoint };
+    LightInfo* info;
```

Comment:
After looking into things more, it appears that Lights_Update and Lights_Draw will do so for every light currently in the node buffer, even if those lights werent added by the system currently calling this function.
To put that another way, Kankyo and Actor lights will still get drawn if you nop their respective functions, but leave the last one in the chain in EfectsSs intact..

Should this instead be UpdateAll and DrawAll, or should that be left to an implementation detail that I can mention in a comment?

Because of this fact, im very confused why the buffer can have up to 32 nodes. It seems like the hard limitation is 7 lights in a LightsN no matter what..

Reply:
I guess to be completely accurate, it would be Lights_UpdateAll which updates every light in the buffer and adds all of them to a LightsN group, and then Lights_Draw draws all of the lights in the LightsN (which again, has a hard limit of 7 lights at a given time for the whole game).

imo seeing Lights_UpdateAll followed by Lights_Draw would be pretty confusing. So Im not sure which naming scheme is better to use for the sake of clarity outside of the lighting system. 

No matter what Ill write comments explaining all of these details. But want help with the naming scheme that will be used externally